fairlearn/reductions/_moments/cdf_demographic_parity_moment.py

```diff
@@ -0,0 +1,198 @@
+                    "Some of the grids do not fall into the range of y"
+            self.grids = np.unique(self.grids)
+            # ensure grids are listed in an increasing order and are unique
+
+        # in case the grids have already included the two endpoints of the range
+        # prepare for the calculation of optimal labels
+        self.pred_grid = [self.y_range[0]] +\
+            list([self.next_grid(grid) for grid in self.grids]) + [self.y_range[1]]
+        self.pred_grid = list(filter(lambda x: x >= self.y_range[0], self.pred_grid))
+        self.pred_grid = list(filter(lambda x: x <= self.y_range[1], self.pred_grid))
+        self.pred_vec = {}
+        for pred in self.pred_grid:
+            self.pred_vec[pred] = (1 * (pred >= pd.Series(self.grids)))
+
+    def prev_grid(self, theta):
+        index = np.where(self.grids == theta)[0][0]
+        if index == 0:
+            return self.grids[index] - (self.grids[1]-self.grids[0])/2
+        else:
+            return (self.grids[index] + self.grids[index-1])/2
+
+    def next_grid(self, theta):
+        index = np.where(self.grids == theta)[0][0]
+        if index + 1 == self.grids.size:
+            return self.grids[index] + (self.grids[1]-self.grids[0])/2
+        else:
+            return (self.grids[index+1] + self.grids[index])/2
+
+    def augment_data(self, x, a, y):
+        """For each x, we create self.grid_num copies and add a new column named 
+        "theta" for each copy. The values of this columns for copies is the chose 
+        grids. 
+
+        The weight for each new data point is calculated in the section 4.1 of 
+        'Agarwal et al. (2019) <https://arxiv.org/abs/1905.12843v1>', but consider
+        the more general case where the grids are not equally spaced.
+
+        The binary label for each new data point is determined by whether the 
+        corresponding weight is positve or not.
+
+
+        Parameters
+        ----------
+        X : numpy.ndarray or pandas.DataFrame
+            Feature data
+        y : numpy.ndarray, pandas.DataFrame, pandas.Series, or list
+            Label vector
+
+        
+        Returns
+        -------
+        pandas.DataFrame, pandas.DataFrame, pandas.DataFrame, pandas.DataFrame
+            Represent the augmented features, sensitive features, labels, and weights.
+        """
+        self.n = np.shape(x)[0]
+        self.width = self.grids[1] - self.grids[0]
+
+        X_aug = pd.concat(repeat(x, self.grid_num))
+        A_aug = pd.concat(repeat(a, self.grid_num))
+        Y_values = pd.concat(repeat(y, self.grid_num))
+
+        theta_list = [s for theta in self.grids for s in repeat(theta, self.n)]
+        X_aug['theta'] = pd.Series(theta_list, index=X_aug.index)
+
+        X_aug.index = range(self.n * self.grid_num)
+        A_aug.index = range(self.n * self.grid_num)
+        Y_values.index = range(self.n * self.grid_num)
+
+        def weight_assign(theta, y):
```

Comment:
are we defining this just to be used once? If so, maybe it's easier not to use a function and just use the logic directly.